File: compiler/bench/permute.fs
\ expected: 3628800
\ Generate all permutations of 10 elements

create perm 10 allot
create used 10 allot
variable perm-count

: permute ( depth -- ) recursive
  dup 10 = if drop 1 perm-count +! exit then
  10 0 do
    i used + c@ 0= if
      i over perm + c!
      1 i used + c!
      dup 1+ recurse
      0 i used + c!
    then
  loop drop ;

: main
  10 0 do 0 i used + c! loop
  0 perm-count !
  0 permute
  perm-count @ . cr ;

File: compiler/bench/permute.c
// expected: 3628800
#include <stdio.h>
int perm[10];
int used[10];
long perm_count = 0;

void permute(int depth) {
    if (depth == 10) { perm_count++; return; }
    for (int i = 0; i < 10; i++) {
        if (!used[i]) {
            perm[depth] = i;
            used[i] = 1;
            permute(depth + 1);
            used[i] = 0;
        }
    }
}

int main() {
    for (int i = 0; i < 10; i++) used[i] = 0;
    permute(0);
    printf("%ld\n", perm_count);
    return 0;
}
